stan/math/prim/prob/beta_neg_binomial_cdf.hpp

```diff
@@ -0,0 +1,168 @@
+                              decltype(b_plus_n)>;
+    auto F = hypergeometric_3F2(
+        std::initializer_list<a_t>{1.0, b_plus_n + 1.0, r_plus_n + 1.0},
+        std::initializer_list<b_t>{n_dbl + 2.0, a_plus_r + b_plus_n + 1.0},
+        1.0);
+    auto C = lgamma(r_plus_n + 1.0) + lbeta(a_plus_r, b_plus_n + 1.0)
+             - lgamma(r_dbl) - lbeta(alpha_dbl, beta_dbl) - lgamma(n_dbl + 2.0);
+    auto ccdf = stan::math::exp(C) * F;
+    cdf *= 1.0 - ccdf;
+
+    if constexpr (!is_constant_all<T_r, T_alpha, T_beta>::value) {
+      auto chain_rule_term = -ccdf / (1.0 - ccdf);
+      auto digamma_n_r_alpha_beta = digamma(a_plus_r + b_plus_n + 1.0);
+      T_partials_return dF[6];
+      grad_F32<false, !is_constant<T_beta>::value, !is_constant_all<T_r>::value,
+               false, true, false>(dF, 1.0, b_plus_n + 1.0, r_plus_n + 1.0,
+                                   n_dbl + 2.0, a_plus_r + b_plus_n + 1.0, 1.0,
+                                   precision, max_steps);
+
+      if constexpr (!is_constant<T_r>::value || !is_constant<T_alpha>::value) {
+        auto digamma_r_alpha = digamma(a_plus_r);
+        if constexpr (!is_constant<T_r>::value) {
+          auto partial_lccdf = digamma(r_plus_n + 1.0)
+                               + (digamma_r_alpha - digamma_n_r_alpha_beta)
+                               + (dF[2] + dF[4]) / F - digamma(r_dbl);
+          partials<0>(ops_partials)[i] += partial_lccdf * chain_rule_term;
+        }
+        if constexpr (!is_constant<T_alpha>::value) {
+          auto partial_lccdf = digamma_r_alpha - digamma_n_r_alpha_beta
+                               + dF[4] / F - digamma(alpha_dbl);
+          partials<1>(ops_partials)[i] += partial_lccdf * chain_rule_term;
+        }
+      }
+
+      if constexpr (!is_constant<T_alpha>::value
+                    || !is_constant<T_beta>::value) {
+        auto digamma_alpha_beta = digamma(alpha_dbl + beta_dbl);
+        if constexpr (!is_constant<T_alpha>::value) {
+          partials<1>(ops_partials)[i] += digamma_alpha_beta * chain_rule_term;
+        }
+        if constexpr (!is_constant<T_beta>::value) {
+          auto partial_lccdf = digamma(b_plus_n + 1.0) - digamma_n_r_alpha_beta
+                               + (dF[1] + dF[4]) / F
+                               - (digamma(beta_dbl) - digamma_alpha_beta);
+          partials<2>(ops_partials)[i] += partial_lccdf * chain_rule_term;
+        }
+      }
+    }
+  }
+
+  if constexpr (!is_constant<T_r>::value) {
+    for (size_t i = 0; i < stan::math::size(r); ++i) {
+      partials<0>(ops_partials)[i] *= cdf;
+    }
+  }
+  if constexpr (!is_constant<T_alpha>::value) {
+    for (size_t i = 0; i < stan::math::size(alpha); ++i) {
+      partials<1>(ops_partials)[i] *= cdf;
+    }
+  }
+  if constexpr (!is_constant<T_beta>::value) {
+    for (size_t i = 0; i < stan::math::size(beta); ++i) {
+      partials<2>(ops_partials)[i] *= cdf;
+    }
+  }
```

Comment:
The value of `cdf` changes once per each iteration of the main loop. Here you add the last calculated cdf value to the entire partial of r, alpha, beta. Do you mean to do this inside of the loop so `partial[i]` gets multiplied by the cdf for that iteration?

Reply:
This is intentionally done after all cdfs have been calculated, in order to comply with the chain rule of multiplication. 
For example, `cdf([entry1,entry2,entry3]) = cdf(entry1) * cdf(entry2) * cdf(entry3)`. When taking derivatives to parameter alpha in `entry1`, one have to times `cdf(entry2) * cdf(entry3)`. This is done by first divide by `cdf(entry1)` for that iteration then times the multiplication of all cdfs. Almost all `*_cdf` in stan have these lines.

Reply:
Oh! I missed the `*=` on the cdf! Thought it was just =